From: Matt Fleming <matt.fleming@intel.com>
Date: Fri, 4 Oct 2013 09:36:56 +0100
Patch-Mainline: v3.13
References: bnc#845706
Subject: [PATCH] x86/efi: Add EFI framebuffer earlyprintk support

It's incredibly difficult to diagnose early EFI boot issues without
special hardware because earlyprintk=vga doesn't work on EFI systems.

Add support for writing to the EFI framebuffer, via earlyprintk=efi,
which will actually give users a chance of providing debug output.

Cc: H. Peter Anvin <hpa@zytor.com>
Acked-by: Ingo Molnar <mingo@kernel.org>
Cc: Thomas Gleixner <tglx@linutronix.de>
Cc: Peter Jones <pjones@redhat.com>
Signed-off-by: Matt Fleming <matt.fleming@intel.com>
Signed-off-by: Oliver Neukum <oneukum@suse.de>
Automatically created from "patches.drivers/0001-x86-efi-Add-EFI-framebuffer-earlyprintk-support.patch" by xen-port-patches.py

--- sle12.orig/arch/x86/kernel/early_printk-xen.c	2013-05-24 11:49:51.000000000 +0200
+++ sle12/arch/x86/kernel/early_printk-xen.c	2014-01-09 12:11:39.000000000 +0100
@@ -16,6 +16,7 @@
 #include <asm/mrst.h>
 #include <asm/pgtable.h>
 #include <linux/usb/ehci_def.h>
+#include <linux/efi.h>
 
 #ifndef CONFIG_XEN
 /* Simple VGA output */
@@ -243,7 +244,8 @@ static int __init setup_early_printk(cha
 			max_ypos = boot_params.screen_info.orig_video_lines;
 			current_ypos = boot_params.screen_info.orig_y;
 #else
-		if (!strncmp(buf, "vga", 3) || !strncmp(buf, "xen", 3)) {
+		if (!strncmp(buf, efi_enabled(EFI_BOOT) ? "efi" : "vga", 3)
+		    || !strncmp(buf, "xen", 3)) {
 #endif
 			early_console_register(&early_vga_console, keep);
 		}
